File: apps/cargo-tracker/src/contexts/handling/domain/model/handling-activity.ts
import { Location } from '../../../../shared/domain/model/location.js';
import { HandlingValidationError } from './handling-validation-error.js';
import { HandlingType } from './handling-type.js';
import { HandlingVoyageNumber } from './handling-voyage-number.js';
import type { CargoSnapshot } from './cargo-snapshot.js';

export interface RegisterHandlingActivityParams {
  bookingId: string;
  type: string;
  location: string;
  completionTime: Date;
  voyageNumber?: string | null;
  operatorName?: string | null;
  consigneeConfirmation?: string | null;
}

/**
 * 荷役作業（集約ルート）。荷役作業の登録と妥当性検証を担う。
 * 妥当性は CargoSnapshot（ACL 経由の貨物情報）に対するデシジョンテーブルで判定する:
 *
 * | 種別    | Voyage 必須 | 場所チェック                     | 不一致時       |
 * | RECEIVE | 不要        | 出発港（origin）と一致           | 警告           |
 * | LOAD    | 必須        | Leg.loadLocation・航海と一致     | MISROUTED      |
 * | UNLOAD  | 必須        | Leg.unloadLocation・航海と一致   | MISROUTED      |
 * | CLAIM   | 不要        | 目的港（destination）と一致      | 警告           |
 * | CUSTOMS | 不要        | なし                             | -              |
 */
export class HandlingActivity {
  private constructor(
    readonly id: number | null,
    readonly bookingId: string,
    readonly type: HandlingType,
    readonly location: Location,
    readonly completionTime: Date,
    readonly voyageNumber: HandlingVoyageNumber | null,
    readonly operatorName: string | null,
    readonly consigneeConfirmation: string | null,
  ) {}

  static register(params: RegisterHandlingActivityParams): HandlingActivity {
    if (params.bookingId.trim().length === 0) {
      throw new HandlingValidationError('予約 ID は必須です');
    }
    if (Number.isNaN(params.completionTime.getTime())) {
      throw new HandlingValidationError('作業日時が不正です');
    }
    const type = HandlingType.of(params.type);
    const voyageNumber =
      params.voyageNumber != null && params.voyageNumber.trim().length > 0
        ? HandlingVoyageNumber.of(params.voyageNumber)
        : null;
    if (type.requiresVoyageNumber() && voyageNumber === null) {
      throw new HandlingValidationError(`${type.value} には航海番号が必要です`);
    }
    // 荷受人確認は引き渡し証明のため CLAIM（引取）でのみ保持する。
    const consigneeConfirmation = type.isClaimType()
      ? (params.consigneeConfirmation ?? null)
      : null;
    return new HandlingActivity(
      null,
      params.bookingId,
      type,
      Location.of(params.location),
      params.completionTime,
      voyageNumber,
      params.operatorName ?? null,
      consigneeConfirmation,
    );
  }

  static reconstruct(params: RegisterHandlingActivityParams & { id: number }): HandlingActivity {
    const activity = HandlingActivity.register(params);
    return new HandlingActivity(
      params.id,
      activity.bookingId,
      activity.type,
      activity.location,
      activity.completionTime,
      activity.voyageNumber,
      activity.operatorName,
      activity.consigneeConfirmation,
    );
  }

  /** 荷役妥当性検証。false のとき、type.misroutesOnMismatch() なら MISROUTED、そうでなければ警告 */
  isValidFor(snapshot: CargoSnapshot): boolean {
    switch (this.type.value) {
      case 'RECEIVE':
        return this.location.unlocode === snapshot.origin;
      case 'LOAD':
        return snapshot.itineraryLegs.some(
          (leg) =>
            leg.loadLocation === this.location.unlocode && leg.voyageNumber === this.voyageNumber?.value,
        );
      case 'UNLOAD':
        return snapshot.itineraryLegs.some(
          (leg) =>
            leg.unloadLocation === this.location.unlocode && leg.voyageNumber === this.voyageNumber?.value,
        );
      case 'CLAIM':
        return this.location.unlocode === snapshot.destination;
      case 'CUSTOMS':
        return true;
    }
  }
}

File: docs/article/source/go/apps/cargo-tracker/internal/handling/domain/handling_activity.go
package domain

import (
	"errors"
	"strings"
	"time"

	shared "github.com/k2works/case-study-cargo-tracker/apps/cargo-tracker/internal/shared/domain"
)

// ドメインエラー。
var (
	ErrEmptyCargoBookingId           = errors.New("cargo booking id must not be empty")
	ErrZeroLocation                  = errors.New("handling location must not be empty")
	ErrVoyageNumberRequired          = errors.New("voyage number is required for load/unload")
	ErrConsigneeConfirmationRequired = errors.New("consignee confirmation is required for claim")
)

// LegSnapshot は CargoSnapshot 内の旅程区間（積込港・荷降港・航海番号）を表す値オブジェクト。
type LegSnapshot struct {
	loadLocation   shared.Location
	unloadLocation shared.Location
	voyageNumber   string
}

// NewLegSnapshot は旅程区間スナップショットを生成する。
func NewLegSnapshot(load, unload shared.Location, voyageNumber string) LegSnapshot {
	return LegSnapshot{loadLocation: load, unloadLocation: unload, voyageNumber: voyageNumber}
}

// LoadLocation は積込港を返す。
func (l LegSnapshot) LoadLocation() shared.Location { return l.loadLocation }

// UnloadLocation は荷降港を返す。
func (l LegSnapshot) UnloadLocation() shared.Location { return l.unloadLocation }

// VoyageNumber は航海番号を返す。
func (l LegSnapshot) VoyageNumber() string { return l.voyageNumber }

// CargoSnapshot は ACL 経由で取得した貨物情報の不変スナップショット。荷役妥当性検証に用いる。
type CargoSnapshot struct {
	bookingId     string
	origin        shared.Location
	destination   shared.Location
	legs          []LegSnapshot
	routingStatus shared.RoutingStatus
}

// NewCargoSnapshot は貨物スナップショットを生成する。
func NewCargoSnapshot(bookingId string, origin, destination shared.Location, legs []LegSnapshot, routingStatus shared.RoutingStatus) CargoSnapshot {
	cp := make([]LegSnapshot, len(legs))
	copy(cp, legs)
	return CargoSnapshot{bookingId: bookingId, origin: origin, destination: destination, legs: cp, routingStatus: routingStatus}
}

// BookingId は予約 ID を返す。
func (s CargoSnapshot) BookingId() string { return s.bookingId }

// Origin は出発港を返す。
func (s CargoSnapshot) Origin() shared.Location { return s.origin }

// Destination は目的港を返す。
func (s CargoSnapshot) Destination() shared.Location { return s.destination }

// Legs は旅程区間を返す。
func (s CargoSnapshot) Legs() []LegSnapshot {
	cp := make([]LegSnapshot, len(s.legs))
	copy(cp, s.legs)
	return cp
}

// ValidationResult は荷役妥当性検証の結果を表す。
type ValidationResult struct {
	matched   bool
	misrouted bool
	warning   bool
}

// Matched は荷役場所が旅程・仕様と一致したかを返す。
func (r ValidationResult) Matched() bool { return r.matched }

// Misrouted は経路不整合（MISROUTED 確定）かを返す。
func (r ValidationResult) Misrouted() bool { return r.misrouted }

// Warning は警告（不一致だが MISROUTED ではない）かを返す。
func (r ValidationResult) Warning() bool { return r.warning }

// HandlingActivity は荷役作業を表す集約ルート。
type HandlingActivity struct {
	cargoBookingId        string
	handlingType          HandlingType
	location              shared.Location
	completionTime        time.Time
	voyageNumber          string
	consigneeConfirmation string
	operatorName          string
}

// NewHandlingActivity はバリデーション付きで荷役作業を生成する。
// LOAD / UNLOAD は voyageNumber 必須、CLAIM は consigneeConfirmation（署名/確認コード）必須。
func NewHandlingActivity(
	cargoBookingId string,
	handlingType HandlingType,
	location shared.Location,
	completionTime time.Time,
	voyageNumber string,
	consigneeConfirmation string,
	operatorName string,
) (HandlingActivity, error) {
	if strings.TrimSpace(cargoBookingId) == "" {
		return HandlingActivity{}, ErrEmptyCargoBookingId
	}
	if location.IsZero() {
		return HandlingActivity{}, ErrZeroLocation
	}
	if handlingType.RequiresVoyageNumber() && strings.TrimSpace(voyageNumber) == "" {
		return HandlingActivity{}, ErrVoyageNumberRequired
	}
	if handlingType.IsClaimType() && strings.TrimSpace(consigneeConfirmation) == "" {
		return HandlingActivity{}, ErrConsigneeConfirmationRequired
	}
	return HandlingActivity{
		cargoBookingId:        cargoBookingId,
		handlingType:          handlingType,
		location:              location,
		completionTime:        completionTime,
		voyageNumber:          voyageNumber,
		consigneeConfirmation: consigneeConfirmation,
		operatorName:          operatorName,
	}, nil
}

// IsValidFor は貨物スナップショットに対する荷役妥当性検証（デシジョンテーブル）を行う。
//
//	RECEIVE: 出発港一致。不一致で警告
//	LOAD:    Leg.loadLocation 一致。不一致で MISROUTED
//	UNLOAD:  Leg.unloadLocation 一致。不一致で MISROUTED
//	CLAIM:   目的港一致。不一致で警告
//	CUSTOMS: 場所チェックなし（常に妥当）
func (a HandlingActivity) IsValidFor(snapshot CargoSnapshot) ValidationResult {
	switch a.handlingType {
	case HandlingTypeReceive:
		return warnResult(a.location.SameAs(snapshot.origin))
	case HandlingTypeClaim:
		return warnResult(a.location.SameAs(snapshot.destination))
	case HandlingTypeLoad:
		return misrouteResult(a.matchesAnyLeg(snapshot, func(l LegSnapshot) shared.Location { return l.loadLocation }))
	case HandlingTypeUnload:
		return misrouteResult(a.matchesAnyLeg(snapshot, func(l LegSnapshot) shared.Location { return l.unloadLocation }))
	default: // CUSTOMS
		return ValidationResult{matched: true}
	}
}

func (a HandlingActivity) matchesAnyLeg(snapshot CargoSnapshot, pick func(LegSnapshot) shared.Location) bool {
	for _, leg := range snapshot.legs {
		if a.location.SameAs(pick(leg)) {
			return true
		}
	}
	return false
}

func warnResult(matched bool) ValidationResult {
	return ValidationResult{matched: matched, warning: !matched}
}

func misrouteResult(matched bool) ValidationResult {
	return ValidationResult{matched: matched, misrouted: !matched}
}

// CargoBookingId は貨物予約 ID を返す。
func (a HandlingActivity) CargoBookingId() string { return a.cargoBookingId }

// Type は荷役種別を返す。
func (a HandlingActivity) Type() HandlingType { return a.handlingType }

// Location は荷役場所を返す。
func (a HandlingActivity) Location() shared.Location { return a.location }

// CompletionTime は作業完了時刻を返す。
func (a HandlingActivity) CompletionTime() time.Time { return a.completionTime }

// VoyageNumber は航海番号を返す。
func (a HandlingActivity) VoyageNumber() string { return a.voyageNumber }

// ConsigneeConfirmation は荷受人確認（署名/確認コード）を返す。
func (a HandlingActivity) ConsigneeConfirmation() string { return a.consigneeConfirmation }

// OperatorName は作業員名を返す。
func (a HandlingActivity) OperatorName() string { return a.operatorName }

// ResultingTransportStatus は荷役完了後に遷移する輸送状態を返す。
func (a HandlingActivity) ResultingTransportStatus() shared.TransportStatus {
	return a.handlingType.ResultingTransportStatus()
}
